tests : timer_api : Prevent compiler optimization

yingmingx opened the pull request:
The local variable 'interval' in timer's callback function
should not be optimized by compiler.
fix #21813

Signed-off-by: Ying ming <mingx.ying@intel.com>

tyrantlizard77 reviewed (commented):
what problem does making this 'volatile' solve?

it's not something that gets modified by another thread

yingmingx commented:
> what problem does making this 'volatile' solve?
> 
> it's not something that gets modified by another thread

As far as I know ，variable could be changed by three situation:
1、changed by hardware
2、changed by ISR
3、changed by another thread
In this problem, the test case was failed occasionally.  When I changed test case order in file main.c, for example, put the test_timer_k_define in the last parameter in the ztest_test_suite function, it Passed. 
So I think the compiler has do some optimization and it have get the latest data in memory.

alexanderwachter commented on tests/kernel/timer/timer_api/src/main.c:
```diff
@@ -57,7 +57,7 @@ static void init_timer_data(void)
 {
 	/** TESTPOINT: expire function */
-	s64_t interval = k_uptime_delta(&tdata.timestamp);
+	volatile s64_t interval = k_uptime_delta(&tdata.timestamp);
```
The scope of this variable ends at line 72 right? Or am I missing something?
It should be totally fine if the compile does optimization here.

Commit:
tests : timer_api : Prevent compiler optimization

The local variable 'interval' in timer's callback function
should not be optimized by compiler.

Signed-off-by: Ying ming <mingx.ying@intel.com>

tyrantlizard77 reviewed (changes requested):
Your change doesn't do what you think it does

andrewboie commented:
If you want to get this merged, you need to show more evidence. Like was done here: https://github.com/zephyrproject-rtos/zephyr/pull/22059

andyross commented:
As mentioned, this change looks to be  a noop.  Check your build/zephyr/zephyr.lst output, I'd bet that there's no difference in the generated code at all.  Or if there is, it's possible there's a compiler bug affecting the test and we should be looking at that more carefully.

If you're getting spurious failures on a hardware platform due to this particular interval being wrong, then that's likely a bug somewhere else with the k_timer callback firing prematurely.  Or the system time going backwards: the SysTick driver is subject to a rollover condition because the 24 bit counter reaches maximum on the FRDM board (and other fast boards) several times per second.  There's some subtle code to manage this, and I wouldn't be shocked if there was a bug in there.  Though it would be showing problems across all the timing tests if this were so.

dleach02 commented:
there is a difference between output but it isn't consequential. In the current state, the inline function k_uptime_delta() has two local variables including 'delta' which, in my build, the value gets retained in r6 after the maths to do the delta and store back time 'uptime' back into the reference time variable. Later in the code when the assert tests are generated it will to the comparison to the value in r6. So basically the compiler optimized out dealing with a stack variable 'interval'

With this change of setting 'volatile' on 'interval', k_uptime_delta() ends up storing 'delta' in r6 and then stores it into the the stack defined variable 'interval'. Later in the code when the assert tests do the compare, the code generated will read out the value stored into the stack into r2 and do the comparison with r2. 

Both are same except there is an overhead in the 'volatile' case of storing/loading the delta into the stack variable 'interval'.

dleach02 reviewed (changes requested):
change doesn't do anything

yingmingx commented:
> change doesn't do anything



> change doesn't do anything
